### ISSUE TITLE
Conflicting definitions for fstat, lstat, stat and fstatat

### ISSUE THREAD

**Issue by catull** [att/ast#248]
Under macOS 10.12.6, src/lib/libast/ast@sha/comp_strtold.c does not compile anymore.

In an unmodified workspace, with the latest sources pulled, the file mentioned produces the same output with clang (stock version, /usr/bin/clang) or gcc (7.1 in this case):

```c
[10/542] Compiling C object 'src/lib/libast/ast@sha/comp_strtold.c.o'.
FAILED: src/lib/libast/ast@sha/comp_strtold.c.o
clang  -Isrc/lib/libast/ast@sha -Isrc/lib/libast -I../src/lib/libast -Isrc/lib/libast/tm -I../src/lib/libast/tm -Isrc/lib/libast/vmalloc -I../src/lib/libast/vmalloc -I../src/lib/libast/std -Isrc/lib/libast/misc -I../src/lib/libast/misc -Isrc/lib/libast/string -I../src/lib/libast/string -Isrc/lib/libast/port -I../src/lib/libast/port -Isrc/lib/libast/path -I../src/lib/libast/path -Isrc/lib/libast/sfio -I../src/lib/libast/sfio -Isrc/lib/libast/comp -I../src/lib/libast/comp -Isrc/lib/libast/cdt -I../src/lib/libast/cdt -Isrc/lib/libast/aso -I../src/lib/libast/aso -I../src/lib/libast/features -I../src/lib/libast/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -O0 -g -D_PACKAGE_ast -DMESON_BUILD=1 -Wno-unknown-pragmas -g '-DHOSTTYPE="darwin.i386-64"' -D_BLD_ast -D_API_ast=20130625 '-D_locale_utf8_str="UTF-8"' '-DCONF_LIBPATH="lib:DYLD_LIBRARY_PATH"' -DALIGN_BOUND2=16 -DALIGN_BOUND1=ALIGN_BOUND2 -DALIGN_BOUND=ALIGN_BOUND2 -Disnanl=isnan -DSTAT_ST_MTIMESPEC=1 -D_hdr_mach_o_dyld -D_ptr_bits=64 -D_more_void_int=1 -D_more_long_int=1 -D_lib_posix_spawn=1 -MMD -MQ 'src/lib/libast/ast@sha/comp_strtold.c.o' -MF 'src/lib/libast/ast@sha/comp_strtold.c.o.d' -o 'src/lib/libast/ast@sha/comp_strtold.c.o' -c ../src/lib/libast/comp/strtold.c
In file included from ../src/lib/libast/comp/strtold.c:50:
In file included from ../src/lib/libast/include/ast.h:34:
In file included from ../src/lib/libast/include/ast_std.h:290:
../src/lib/libast/include/ast_intercept.h:212:35: warning: declaration of 'struct stat' will not be visible outside of this function [-Wvisibility]
extern int              ast_fstat(int, struct stat*);
                                              ^
../src/lib/libast/include/ast_intercept.h:227:42: warning: declaration of 'struct stat' will not be visible outside of this function [-Wvisibility]
extern int              ast_stat(const char*, struct stat*);
                                                     ^
../src/lib/libast/include/ast_intercept.h:233:50: warning: declaration of 'struct stat' will not be visible outside of this function [-Wvisibility]
extern int              ast_fstatat(int, const char*, struct stat*, int);
                                                             ^
../src/lib/libast/include/ast_intercept.h:235:43: warning: declaration of 'struct stat' will not be visible outside of this function [-Wvisibility]
extern int              ast_lstat(const char*, struct stat*);
                                                      ^
In file included from ../src/lib/libast/comp/strtold.c:52:
In file included from ../src/lib/libast/sfio/sfstrtof.h:48:
In file included from ../src/lib/libast/sfio/sfhdr.h:83:
In file included from ../src/lib/libast/include/ls.h:34:
/usr/include/sys/stat.h:354:5: error: conflicting types for 'ast_fstat'
int     fstat(int, struct stat *) __DARWIN_INODE64(fstat);
        ^
../src/lib/libast/include/ast_intercept.h:126:21: note: expanded from macro 'fstat'
#define fstat(a,b)              ast_fstat(a,b)
                                ^
../src/lib/libast/include/ast_intercept.h:212:13: note: previous declaration is here
extern int              ast_fstat(int, struct stat*);
                        ^
In file included from ../src/lib/libast/comp/strtold.c:52:
In file included from ../src/lib/libast/sfio/sfstrtof.h:48:
In file included from ../src/lib/libast/sfio/sfhdr.h:83:
In file included from ../src/lib/libast/include/ls.h:34:
/usr/include/sys/stat.h:355:5: error: conflicting types for 'ast_lstat'
int     lstat(const char *, struct stat *) __DARWIN_INODE64(lstat);
        ^
../src/lib/libast/include/ast_intercept.h:143:16: note: expanded from macro 'lstat'
#define lstat           ast_lstat
                        ^
../src/lib/libast/include/ast_intercept.h:235:13: note: previous declaration is here
extern int              ast_lstat(const char*, struct stat*);
                        ^
In file included from ../src/lib/libast/comp/strtold.c:52:
In file included from ../src/lib/libast/sfio/sfstrtof.h:48:
In file included from ../src/lib/libast/sfio/sfhdr.h:83:
In file included from ../src/lib/libast/include/ls.h:34:
/usr/include/sys/stat.h:358:5: error: conflicting types for 'ast_stat'
int     stat(const char *, struct stat *) __DARWIN_INODE64(stat);
        ^
../src/lib/libast/include/ast_intercept.h:178:19: note: expanded from macro 'stat'
#define stat(p,s)       ast_stat(p,s)
                        ^
../src/lib/libast/include/ast_intercept.h:227:13: note: previous declaration is here
extern int              ast_stat(const char*, struct stat*);
                        ^
In file included from ../src/lib/libast/comp/strtold.c:52:
In file included from ../src/lib/libast/sfio/sfstrtof.h:48:
In file included from ../src/lib/libast/sfio/sfhdr.h:83:
In file included from ../src/lib/libast/include/ls.h:34:
/usr/include/sys/stat.h:364:5: error: conflicting types for 'ast_fstatat'
int     fstatat(int, const char *, struct stat *, int) __DARWIN_INODE64(fstatat) __OSX_AVAILABLE_STARTING(__MAC_10_10, __IPHONE_8_0);
        ^
../src/lib/libast/include/ast_intercept.h:59:18: note: expanded from macro 'fstatat'
#define fstatat         ast_fstatat
                        ^
../src/lib/libast/include/ast_intercept.h:233:13: note: previous declaration is here
extern int              ast_fstatat(int, const char*, struct stat*, int);
```

**Comment by krader1961** [comment]
This was introduced by the most recent set of `iffe` simplifications by @siteshwar. I would have caught it but forgot to do the `git clean -dfx` dance that is needed when making that type of change. So on my server it was still using the old headers. I'll clean this up since I let it pass through.

Note that related to this the new static headers assume `getmntent()` is available (along with related header definitions). Which isn't true on macOS or FreeBSD. I just merged a commit to fix that build failure. Since the `ksh` command doesn't need that code I just removed the modules that need those features and updated the Meson config.
